### PR TITLE
doc: Update cloud-init.md with user-data examples

### PR DESCRIPTION
**Pull request by cricalix**
Context
==
I found the current documentation to be very lacking in terms of a clear explanation of how to apply the `user-data` configuration options, and how to confirm that the `cloud-init` process had received the data. It was only through a lot of searching that I found out that `#cloud-config` is actually not just a comment, but actually required for `cloud-init` to even process that set of piped configuration keys.

Approach
==
This change
* Adds a small "how to" for setting up a new profile for testing
* Adds documentation on the pipe syntax that's needed to pipe the configuration to cloud-init
* Adds documentation for the specific syntax required to get `cloud-init.user-data` to work (`#cloud-config`)
* Adds documentation on where to check on the instance to see if the user-data was passed in properly
* Adds examples of using the `user-data` section to take actions in an instance

Errata
==
I acknowledge that v5 is meant to bring a documentation rewrite per the comment on the YouTube video linked from this document, but I don't think that's a good reason to not have some clear examples in the documentation right now. Google searches tend (for me) to find documentation from 2018 that use the old config syntax (user.user-data), and that documentation doesn't say `#cloud-config` is crucial.

Testing of instructions
==

With the `#cloud-config` line in the `cloud-init.user-data` block:
```
$ ls -l /var/lib/cloud/instance | grep -E 'cloud-config.txt|user-data.txt'
-rw------- 1 root root  149 Jul 16 08:47 cloud-config.txt
-rw------- 1 root root  117 Jul 16 08:47 user-data.txt
```
`/var/log/syslog` confirms `apt upgrade` and `apt install` were run.

Without the `#cloud-config` line:
```
$ ls -l /var/lib/cloud/instance | grep -E 'cloud-config.txt|user-data.txt'
-rw------- 1 root root    0 Jul 16 08:09 cloud-config.txt
-rw------- 1 root root  103 Jul 16 08:09 user-data.txt
```

`/var/log/syslog` confirms that none of the upgrades etcetera ran.

Signed-off-by: Duncan Hill <github.com@cricalix.net>

### CODE REVIEW

File: doc/cloud-init.md [hunk 1]
```diff
@@ -31,6 +31,111 @@ For LXD instances, `vendor-data` should be used in profiles rather than the inst
...
+    path: /
+    pool: default
+    type: disk
+```
+
+Once you've set up the `cloud-init` configuration, use `lxc launch` with `--profile <profilename>` to apply the profile to the instance.
+
+## Adding cloud-init keys to the configuration
+
+cloud-init keys require a specific syntax. You use a pipe symbol (`|`) to indicate that LXD should pipe the contents of the relevant key
```
Comment (stgraber) [inline doc/cloud-init.md:67]: That is no what the `|` does in YAML. In YAML it simply means that the key content will be on following indented lines.

File: doc/cloud-init.md [hunk 2]
```diff
@@ -31,6 +31,111 @@ For LXD instances, `vendor-data` should be used in profiles rather than the inst
...
+```yaml
+config:
+  cloud-init.network-config: |
+```
+
+## Custom user-data configuration
+
+cloud-init uses the `user-data` (and `vendor-data`) section to do things like upgrade packages, install packages, or run arbitrary commands.
+
+A `cloud-init.user-data` key must have a first line of `#cloud-config`; without this marker, `cloud-init` will not process the remaining configuration.
```
Comment (stgraber) [inline doc/cloud-init.md:88]: That's incorrect. Cloud-init supports multiple syntaxes, one of which happens to be `#cloud-config`

Comment (stgraber) [reply doc/cloud-init.md:88]: See https://cloudinit.readthedocs.io/en/latest/topics/format.html for all the valid formats.

File: doc/cloud-init.md (hunk 1 above)
Comment (cricalix) [reply doc/cloud-init.md:67]: Thanks. I've never actually worked with YAML. Have now gone and read the YAML spec, and will link it in the re-worked text.

File: doc/cloud-init.md (hunk 2 above)
Comment (cricalix) [reply doc/cloud-init.md:88]: Thanks. Section updated, linking to the formats.

File: doc/cloud-init.md [hunk 3]
```diff
@@ -31,6 +31,120 @@ For LXD instances, `vendor-data` should be used in profiles rather than the inst
 
 Cloud-config examples can be found here: https://cloudinit.readthedocs.io/en/latest/topics/examples.html
 
+# Working with cloud-init
```
Comment (ru-fu) [inline doc/cloud-init.md:34]:
```suggestion
## Working with cloud-init
```
Wrong heading level.

File: doc/cloud-init.md [hunk 4]
```diff
@@ -31,6 +31,120 @@ For LXD instances, `vendor-data` should be used in profiles rather than the inst
...
+    type: nic
+  root:
+    path: /
+    pool: default
+    type: disk
+```
+
+Once you've set up the `cloud-init` configuration, use `lxc launch` with `--profile <profilename>` to apply the profile to the instance.
+
+## Adding cloud-init keys to the configuration
```
Comment (ru-fu) [inline doc/cloud-init.md:65]:
```suggestion
### Adding cloud-init keys to the configuration
```

File: doc/cloud-init.md [hunk 5]
```diff
@@ -31,6 +31,120 @@ For LXD instances, `vendor-data` should be used in profiles rather than the inst
...
+
+cloud-init uses the `user-data` (and `vendor-data`) section to do things like upgrade packages, install packages, or run arbitrary commands.
+
+A `cloud-init.user-data` key must have a first line that indicates what type of [data format](https://cloudinit.readthedocs.io/en/latest/topics/format.html) is being passed to `cloud-init`. For activities like upgrading packages or setting up a user, `#cloud-config` is the data format to use.
+
+An instance's rootfs will contain the following files as a result:
+* /var/lib/cloud/instance/cloud-config.txt
+* /var/lib/cloud/instance/user-data.txt
+
+### Upgrade packages on instance creation
```
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/cloud-init.md:94]: The same change for the remaining headings.

File: doc/cloud-init.md [hunk 6]
```diff
@@ -31,6 +31,120 @@ For LXD instances, `vendor-data` should be used in profiles rather than the inst
...
+config:
+  cloud-init.user-data: |
+    #cloud-config
+    packages:
+      - git
+      - openssh-server
+```
+
+### Set the timezone on instance creation
+To set the timezone for the instance, use the `timezone` key:
```
Comment (ru-fu) [inline doc/cloud-init.md:117]:
```suggestion
### Set the time zone on instance creation
To set the time zone for the instance, use the `timezone` key:
```

**Review by ru-fu — COMMENTED** [review]
Thanks!
A bit of language and format nitpicking, but otherwise this looks good and really helpful. :-)

**Comment by stgraber** [comment]
I'll let @ru-fu do another review, if she's happy with it. I'll do a rebase + merge.

**Review by ru-fu — APPROVED** [review]
Looks good now, thank you!

**Pull request merged**